RetailCoder.VBE/Refactorings/Rename/RenameHandlerBase.cs

```diff
@@ -0,0 +1,106 @@
+    {
+        private readonly RenameModel _model;
+        private readonly IMessageBox _messageBox;
+        private List<QualifiedModuleName> _updatedQMNs;
+
+        public RenameHandlerBase(RenameModel model, IMessageBox messageBox)
+        {
+            _model = model;
+            _messageBox = messageBox;
+            _updatedQMNs = new List<QualifiedModuleName>();
+        }
+
+        public RenameModel Model { get { return _model; } }
+
+        public IMessageBox MessageBox { get { return _messageBox; } }
+
+        public void Rewrite()
+        {
+            _updatedQMNs.Distinct().ToList()
+                .ForEach(qmn => Model.State.GetRewriter(qmn).Rewrite());
+        }
+
+        public virtual void Rename() { }
+        public virtual string ErrorMessage { get; }
+
+        public void RenameUsages(Declaration target)
+        {
+            RenameUsages(target, _model.NewName);
+        }
+
+        public void RenameUsages(Declaration target, string newName)
+        {
+            var qualifiedModuleNames = new List<QualifiedModuleName>();
+            var modules = target.References.GroupBy(r => r.QualifiedModuleName);
+            foreach (var grouping in modules)
+            {
+                qualifiedModuleNames.Add(grouping.Key);
+                var rewriter = _model.State.GetRewriter(grouping.Key);
+                var module = grouping.Key.Component.CodeModule;
+                foreach (var line in grouping.GroupBy(reference => reference.Selection.StartLine))
+                {
+                    var lastSelection = Selection.Empty;
+                    foreach (var reference in line.OrderByDescending(l => l.Selection.StartColumn))
+                    {
+                        if (reference.Selection == lastSelection)
+                        {
+                            continue;
+                        }
+                        var newContent = reference.Context.GetText().Replace(reference.IdentifierName, newName);
+                        rewriter.Replace(reference.Context, newContent);
+                        lastSelection = reference.Selection;
+                    }
+                }
+            }
+            _updatedQMNs.AddRange(qualifiedModuleNames);
+        }
+
+        public void RenameDeclaration(Declaration target)
+        {
+            RenameDeclaration(target, _model.NewName);
+        }
+
+        public void RenameDeclaration(Declaration target, string newName)
+        {
+            var qualifiedModuleName = target.QualifiedName.QualifiedModuleName;
+            var component = qualifiedModuleName.Component;
+            var rewriter = _model.State.GetRewriter(target);
+            var module = component.CodeModule;
+            if (!target.DeclarationType.HasFlag(DeclarationType.Property))
+            {
+                var newContent = target.Context.GetText().Replace(target.IdentifierName, newName);
+                rewriter.Replace(target.Context, newContent);
```

Comment:
Now, the problem is that you have a `ParserRuleContext`, and you don't know what specific concrete type it is; all of these `static` method will throw a `NullReferenceException` if you give them a `null` context, so `as`-casting is out of the question. unless...

````csharp
public static string GetName(VBAParser.SubStmtContext context, out Interval tokenInterval)
{
    if (context == null) { tokenInterval = Interval.Invalid; return null; }
    return GetName(context.subroutineName(), out tokenInterval);
}
````

Then you could call each overload with an `as`-cast, and use the single one that doesn't return `null`. Something like:

    Interval tokenInterval;
    var name = Identifier.GetName(context as VBAParser.SubStmtContext, out tokenInterval)?
                         .GetName(context as VBAParser.FunctionStmtContext, out tokenInterval)?
                         .GetName(context as VBAParser.PropertyGetStmtContext, out tokenInterval)?
                         .GetName(context as VBAParser.PropertyLetStmtContext, out tokenInterval)?
                         .GetName(context as VBAParser.PropertySetStmtContext, out tokenInterval)?
                         .GetName(context as VBAParser.WhateverOtherStmtContext, out tokenInterval)?
                        .....